lambda/trmc.ml

```diff
@@ -0,0 +1,515 @@
+*)
+type settable_zipper = {
+  rev_returns : return list;
+  settable : settable;
+  tail_choices: choice list
+}
+let find_settable : choice list -> (settable_zipper, return list) result =
+  let rec find rev_returns = function
+    | [] -> Error (List.rev rev_returns)
+    | Return r :: rest -> find (r :: rev_returns) rest
+    | Set settable :: tail_choices -> Ok { rev_returns; settable; tail_choices }
+  in find []
+
+type context = {
+  candidates: candidate Ident.Map.t;
+}
+and candidate = {
+  arity: int;
+  dps_id: Ident.t;
+}
+
+let set loc dst = function
+  | Return t ->
+      assign_to_dst loc dst t
+  | Set settable ->
+      settable.dps dst
+
+let return = function
+  | Return t -> t
+  | Set settable -> settable.direct ()
+
+let trmc_placeholder = Lconst (Const_base (Const_int 0))
+(* TODO consider using a more magical constant like 42, for debugging? *)
+
+let find_candidate = function
+  | Lfunction lfun when lfun.attr.trmc_candidate -> Some lfun
+  | _ -> None
+
+let declare_binding ctx (var, def) =
+  match find_candidate def with
+  | None -> ctx
+  | Some lfun ->
+  let arity = List.length lfun.params in
+  let dps_id = Ident.create_local (Ident.name var ^ "_trmc") in
+  let cand = { arity; dps_id } in
+  { candidates = Ident.Map.add var cand ctx.candidates }
+
+let rec choice ctx t =
+  let rec choice ctx t =
+    begin[@warning "-8"]
+      (*FIXME: allows non-exhaustive pattern matching;
+        use an overkill functor-based solution instead? *)
+      match t with
+      | (Lvar _ | Lconst _ | Lfunction _ | Lsend _
+        | Lassign _ | Lfor _ | Lwhile _) ->
+          let t = traverse ctx t in
+          Return t
+
+      (* [choice_prim] handles most primitives, but the important case of construction
+         [Lprim(Pmakeblock(...), ...)] is handled by [choice_makeblock] *)
+      | Lprim (prim, primargs, loc) ->
+          choice_prim ctx prim primargs loc
+
+      (* [choice_apply] handles applications, in particular tail-calls which
+         generate Set choices at the leaves *)
+      | Lapply apply ->
+          choice_apply ctx apply
+      (* other cases use the [lift] helper that takes the sub-terms in tail
+         position and the context around them, and generates a choice for
+         the whole term from choices for the tail subterms. *)
+      | Lsequence (l1, l2) ->
+          let l1 = traverse ctx l1 in
+          lift ctx [l2] @@ fun [l2] ->
+          Lsequence (l1, l2)
+      | Lifthenelse (l1, l2, l3) ->
+          let l1 = traverse ctx l1 in
+          lift ctx [l2; l3]
+            (fun [l2; l3] -> Lifthenelse (l1, l2, l3))
+      | Llet (lk, vk, var, def, body) ->
+          (* non-recursive bindings are not specialized *)
+          let def = traverse ctx def in
+          lift ctx [body] @@ fun [body] ->
+          Llet (lk, vk, var, def, body)
+      | Lletrec (bindings, body) ->
+          let ctx, bindings = traverse_letrec ctx bindings in
+          lift ctx [body] @@ fun [body] ->
```

Comment:
Same as below -- the original `ctx` should be used, not the one returned by `traverse_letrec` (otherwise the letrec bindings could be called in destination-passing-style in other letrec bindings inside the body).

Reply:
I propose to leave global propagation for now, and refine this in a later commit.